climakitae/threshold_tools.py

```diff
@@ -884,3 +887,346 @@ def return_period(ams):
+    0 for False, 1 for True, and NaN for NaN values
+    """
+
+    # Count occurances (and preserve NaNs)
+    if threshold_direction == "above":
+        events_da = (da > threshold_value).where(da.isnull()==False)
+    elif threshold_direction == "below":
+        events_da = (da < threshold_value).where(da.isnull()==False)
+    else:
+        raise ValueError(f"Unknown value for `threshold_direction` parameter: {threshold_direction}. Available options are 'above' or 'below'.")
+
+    # Groupby 
+    if groupby is not None:
+        group_totals = _group_and_sum(events_da, groupby)
+        events_da = (group_totals > 0).where(group_totals.isnull()==False)
+
+    return events_da
+
+def _group_and_sum(da, group_spec):
+    """
+    Helper function to sum data across time periods (i.e. days, months years).
+    The `group_spec` argument is a tuple of the form (3, 'day') or (1, 'year').
+
+    Return value is an xarray DataArray, with all the same dimensions except 
+    for `time`, which will be collapsed in length along the groups, and the 
+    coordinate values will be the date of the start of the group, rather than a 
+    datetime. (TBD if this is the desired behavior for the time dimension.)
+
+    This function is implemented by manually creating group numbers (i.e. 
+    [0,0,...1,1,...]), assigning them as a dimension, then summing across 
+    those groups. 
+    """
+    group_len, group_type = group_spec
+
+    if (group_spec == (1, "hour") and da.frequency == "1hr") \
+        or (group_spec == (1, "day") and da.frequency == "1day") \
+        or (group_spec == (1, "month") and da.frequency == "1month"):
+        # group_spec same as data frequency, do nothing
+        group_totals = da
+
+    elif group_spec == (1, "day"):
+        # special case where it is simpler to sum by day using "time.date"
+        group_totals = da.groupby("time.date").sum()
+        group_totals["date"] = pd.to_datetime(group_totals.date)
+        group_totals = group_totals.rename({"date":"time"})
+
+    elif group_type == "day":
+        # general case for grouping by some number of days
+        dates = da.time.dt.date.values # get the date for each value in the time dimension
+        days = [(d - dates[0]).days for d in dates] # calculate the day number for each value (i.e. [0,0,...1,1,...])
+        groups = [math.floor(d / group_len) for d in days] # group the day numbers based on user-specified group length
+        date_ids = dates[[groups.index(i) for i in list(set(groups))]] # save one date value for each group to use as the time index after summing
+        da["time"] = groups # set the group numbers as the time dimension for summing
+        group_totals = da.groupby("time").sum() # sum across the groups
+        group_totals["time"] = pd.to_datetime(date_ids) # reset the time dimension to the saved date values
+
+    else:
+        raise ValueError("Groupby options other than 'day' not yet implmented.")
+
+    return group_totals
```

Comment:
This is the helper function I would like someone to review and possibly discuss with!

Reply:
Aggregating on the temporal dimension can be handled with [.resample()](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.resample.html) which works with various date offset aliases from Pandas and Xarray. You can sum over 3 days:
``` python
da2 = da.resample(time='3D').sum()
```
For annual, `A` and `Y` are synonymous:
``` python
da2 = da.resample(time='A').sum()
```

Reply:
perfect, thanks so much! will work on switching to this instead.